File: Arrays & Strings/7_intersection.java
import java.util.List;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

class Source {
    public static List<Integer> intersection(List<Integer> listA, List<Integer> listB) {
        List<Integer> result = new ArrayList<>();
        Set<Integer> set = new HashSet<>(listA); // O(n)

        // Same as line 9
        // for(int val : listA) {
        // set.add(val);
        // }

        // O(m)
        for (int val : listB) {
            if (set.contains(val)) { // O(1)
                result.add(val);
            }
        }

        return result;
    }

}

File: 1. Arrays & Strings/7_intersection.py
def intersection(list_a, list_b):
    result = []
    set_a = set(list_a)  # O(n)

    # O(m)
    for val in list_b:
        if val in set_a:  # O(1)
            result.append(val)
    
    return result
